crates/turbo-tasks-bytes/src/stream.rs

```diff
@@ -0,0 +1,219 @@
+                source: Box::new(source),
+                pulled: vec![],
+            })),
+        }
+    }
+}
+
+impl<T: Send + Sync + 'static> Stream<T> {
+    /// Constructs a new Stream, and leaves it open for new values to be
+    /// written.
+    pub fn new_open(data: Vec<T>) -> (UnboundedSender<T>, Self) {
+        let (sender, receiver) = unbounded_channel();
+        (
+            sender,
+            Self {
+                inner: Arc::new(Mutex::new(StreamState::OpenStream {
+                    source: Box::new(ReceiverStream { receiver }),
+                    pulled: data,
+                })),
+            },
+        )
+    }
+}
+
+impl<T> Clone for Stream<T> {
+    // The derived Clone impl will only work if `T: Clone`, which is wrong because
+    // we just need to clone the Arc, not the internal data.
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+impl<T> Default for Stream<T> {
+    fn default() -> Self {
+        Self {
+            inner: Arc::new(Mutex::new(StreamState::Closed(Box::new([])))),
+        }
+    }
+}
+
+impl<T: PartialEq> PartialEq for Stream<T> {
+    // A Stream is equal if its the same internal pointer, or both streams are
+    // closed with equivalent values.
+    fn eq(&self, other: &Self) -> bool {
+        Arc::ptr_eq(&self.inner, &other.inner) || {
+            let left = self.inner.lock().unwrap();
+            let right = other.inner.lock().unwrap();
+
+            match (&*left, &*right) {
+                (StreamState::Closed(a), StreamState::Closed(b)) => a == b,
+                _ => false,
+            }
+        }
+    }
+}
+impl<T: Eq> Eq for Stream<T> {}
+
+impl<T: Serialize> Serialize for Stream<T> {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        use serde::ser::Error;
+        let lock = self.inner.lock().map_err(Error::custom)?;
+        match &*lock {
+            StreamState::Closed(data) => data.serialize(serializer),
+            _ => Err(Error::custom("cannot serialize open stream")),
+        }
+    }
+}
+
+impl<'de, T: Deserialize<'de>> Deserialize<'de> for Stream<T> {
+    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        let data = <Vec<T>>::deserialize(deserializer)?;
+        Ok(Stream::new_closed(data))
+    }
+}
+
+impl<T: fmt::Debug> fmt::Debug for StreamState<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::OpenStream { pulled: data, .. } => f
+                .debug_struct("StreamState::OpenStream")
+                .field("data", data)
+                .finish(),
+            Self::Closed(data) => f.debug_tuple("StreamState::Closed").field(data).finish(),
+        }
+    }
+}
+
+/// Implements [StreamTrait] over our Stream.
+#[derive(Debug)]
+pub struct StreamRead<T> {
+    index: usize,
+    source: Stream<T>,
+}
```

Comment:
I think this isn't necessary and can be merged into the `Stream` struct, by adding the `index` field to this struct.